pkg/scheduler/internal/cache/cache.go

```diff
@@ -230,28 +253,66 @@ func (cache *schedulerCache) UpdateNodeInfoSnapshot(nodeSnapshot *nodeinfosnapsh
 	}
 
 	if len(nodeSnapshot.NodeInfoMap) > len(cache.nodes) {
-		for name := range nodeSnapshot.NodeInfoMap {
-			if _, ok := cache.nodes[name]; !ok {
-				delete(nodeSnapshot.NodeInfoMap, name)
-			}
-		}
+		cache.removeDeletedNodesFromSnapshot(nodeSnapshot)
+		updateAllLists = true
 	}
 
-	// Take a snapshot of the nodes order in the tree
+	if updateAllLists || updateNodesHavePodsWithAffinity {
```

Comment:
I think we can optimize further by tracking 1)the list of nodes that are added/removed and 2) list of nodes with affinity pods changed. And in `updateNodeInfoSnapshotList`, we only update the updated nodes instead of recreating the entire node list(s).

But in order to maintain the same order as the node tree, this approach may require the snapshot to maintain a node tree structure. And the node list need to be a linked list for faster insertion/removal.

Need some more thoughts on this.

Reply:
Just to clarify: I am not suggesting to explore this idea in this PR. This PR is good enough!

Reply:
Not clear to me how much we would gain by adding this complexity, certainly nothing in the benchmark test we have (the profile shows that this now is insignificant).

Reply:
it could be useful for clusters that constantly change size, but I think the overhead of bringing up/down nodes is more significant than this.

Reply:
I agree that maintaining nodes added/removed may not be very important. But I think it certainly can help with the cluster where affinity is used a lot.